File: src/Services/ConferenceManagement/PaderConference.Core/Services/Chat/Channels/ChannelSerializer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using PaderConference.Core.Services.Synchronization;

namespace PaderConference.Core.Services.Chat.Channels
{
    public static class ChannelSerializer
    {
        private const string PROP_TYPE = "type";

        public static SynchronizedObjectId Encode(ChatChannel channel)
        {
            var parameters = new Dictionary<string, string>();
            ChatChannelType type;

            switch (channel)
            {
                case GlobalChatChannel:
                    type = ChatChannelType.Global;
                    break;
                case PrivateChatChannel privateChatChannel:
                    type = ChatChannelType.Private;

                    var participants = privateChatChannel.Participants.ToList();
                    parameters.Add("p1", participants[0]);
                    parameters.Add("p2", participants[1]);
                    break;
                case RoomChatChannel roomChatChannel:
                    type = ChatChannelType.Room;
                    parameters.Add("roomId", roomChatChannel.RoomId);
                    break;
                default:
                    throw new ArgumentOutOfRangeException(nameof(channel));
            }

            return CreateSyncChannel(type, parameters);
        }

        public static ChatChannel Decode(SynchronizedObjectId channel)
        {
            var typeString = channel.Parameters[PROP_TYPE];
            var type = StringToChannelType(typeString);

            switch (type)
            {
                case ChatChannelType.Global:
                    return GlobalChatChannel.Instance;
                case ChatChannelType.Room:
                    return new RoomChatChannel(channel.Parameters["roomId"]);
                case ChatChannelType.Private:
                    return new PrivateChatChannel(new HashSet<string>
                    {
                        channel.Parameters["p1"], channel.Parameters["p2"],
                    });
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }

        private static SynchronizedObjectId CreateSyncChannel(ChatChannelType type,
            Dictionary<string, string> additionalParameters)
        {
            additionalParameters.Add(PROP_TYPE, ChannelTypeToString(type));
            return new SynchronizedObjectId(SynchronizedObjectIds.CHAT, additionalParameters);
        }

        private static string ChannelTypeToString(ChatChannelType type)
        {
            return type.ToString().ToLower();
        }

        private static ChatChannelType StringToChannelType(string s)
        {
            return Enum.Parse<ChatChannelType>(s, true);
        }
    }
}

File: src/Web/WebSPA/Client/src/features/chat/channel-serializer.ts
import { parseSynchronizedObjectId } from 'src/store/signal/synchronization/synchronized-object-id';
import { CHAT } from 'src/store/signal/synchronization/synchronized-object-ids';

export type GlobalChatChannel = { type: 'global' };
export type RoomChatChannel = { type: 'room'; roomId: string };
export type PrivateChatChannel = { type: 'private'; participants: string[] };

export type ChatChannel = GlobalChatChannel | RoomChatChannel | PrivateChatChannel;

export type ChatChannelWithId = ChatChannel & { id: string };

export function decode(id: string): ChatChannelWithId {
   const syncObjId = parseSynchronizedObjectId(id);
   if (syncObjId.id !== CHAT) throw new Error('Can only decode a synchronized object that is a chat.');

   const type = syncObjId.parameters['type'];
   switch (type) {
      case 'global':
         return { type, id };
      case 'room':
         return { type, roomId: syncObjId.parameters['roomId'], id };
      case 'private':
         return { type, participants: [syncObjId.parameters['p1'], syncObjId.parameters['p2']], id };
      default:
         throw new Error(`Invalid type: ${type}`);
   }
}
